rfcs/012 - Either, Option, and Result Destructuring.md

```diff
@@ -0,0 +1,110 @@
+
+## Expected SemBDD Impact
+
+If Alan was at least version 1.0.0, then this would be considered a major change that breaks user code when destructuring a `Result<T>`.
+
+## Proposal
+
+This proposal seeks to introduce a special case when handling `if` statements in Alan.
+Currently, when handling a `Result<T>`, a user might expect to write something like this in order to retrieve a value.
+```
+let myResult = ok(5);
+if myResult.isOk() {
+	let okVal = myResult.getOr(-1);
+	print("should be 5: " + okVal);
+} else {
+	let errVal = myResult.getErr("expected error, but could not retrieve Error value");
+	print("unreachable, but I know that errVal is NOT 'expected error, but could not retrieve Error value', unless that's what myResult is");
+}
+```
+
+Note that, in both branches of the above `if` statement, the status (`ok` or `err`) of the `Result` is already known, but the user must then call either `getOr` or `getErr` in order to retrieve the expected value.
+(Note that `getOrDefault` is not yet implemented, but only makes this process marginally easier.)
+Instead, it would be much easier for the user (and potentially even more efficient, although marginally) if the variable could be immediately used as if it were destructured.
+Since `isOk` already guarantees that the `Result` is `ok` (when it returns `true`, otherwise the guarantee is that it's `err`), it would make sense to be able to write the above code like this:
+```
+let myResult = ok(5);
+if myResult.isOk() {
+	print("should be 5: " + myResult); // 1
+} else {
+	print("unreachable, but if myResult is an error, then this would print that error: " + myResult); // 2
+}
+```
+
+In the above code, `myResult` was already destructured into its `ok` value at location 1
+(in other words, at location 1, `myResult` is of type `int64`, not `Result<int64>`).
+This allows the user to quickly and easily perform operations on that inner value without the additional step mentioned above.
+Note that, if myResult were an `err` instead, it would work the exact same way, with the `Error` taking the place of `myResult`.
+
+This should also be extended to `Option<T>` and `Either<T, U>` as well:
+```
+let myOption = some(true);
+if myOption.isSome() {
+	print("here ya go: " + myOption); // 1
+} else {
+	print("I've got nothing for ya"); // 2
+}
+
+let myEither = main("hello, world!"); // 3
+if myEither.isMain() {
+	print("main rock: " + myEither); // 4
+} else {
+	print("alt rock: " + myEither); // 5
+}
+```
+
+In location 1, `myOption` should be of type `bool`.
+Meanwhile, in location 2, it might make sense to undefine the `myOption` var - this is fairly ambiguous and needs more discussion.
+Note that in location 3, `myEither` would technically be assigned the type `Either<string, void>`, but imagine it somehow assigned to a stringable value, say an `int64`.
```

Comment:
You could get the correct type signature if it came from a function that itself was dealing with some conditional, eg:

```
fn over9000(num: int64): Either<string, bool> {
  if num > 9000 {
    return main("It's over 9000!");
  } else {
    return alt(false);
  }
}
...
let myEither = over9000(9001);
```